src/vs/workbench/contrib/mcp/common/mcpService.ts

```diff
@@ -5,83 +5,122 @@
+		@IExtensionService private readonly _extensionService: IExtensionService,
+		@IStorageService storageService: IStorageService,
+		@ILogService private readonly _logService: ILogService,
 	) {
 		super();
 
-		this.userCache = instantiationService.createInstance(McpServerMetadataCache, StorageScope.PROFILE);
-		this.workspaceCache = instantiationService.createInstance(McpServerMetadataCache, StorageScope.WORKSPACE);
-
-		const definitionsObservable = derived(reader => {
-			const collections = this._mcpRegistry.collections.read(reader);
-			return collections.flatMap(collectionDefinition => collectionDefinition.serverDefinitions.read(reader).map(serverDefinition => ({
-				serverDefinition,
-				collectionDefinition,
-			})));
-		});
-
-		const updateThrottle = this._store.add(new RunOnceScheduler(() => {
-			const definitions = definitionsObservable.get();
-
-			const nextDefinitions = new Set(definitions);
-			const currentServers = this._servers.get();
-			const nextServers: IMcpServer[] = [];
-			for (const server of currentServers) {
-				const match = definitions.find(d => defsEqual(server, d));
-				if (match) {
-					nextDefinitions.delete(match);
-					nextServers.push(server);
-				} else {
-					server.dispose();
+		this._register(storageService.onWillSaveState(() => {
+			for (const collectionId of this._extensionCollectionIdsToPersist) {
+				const publishedServers = this._mcpRegistry.collections.get().find(c => c.id === collectionId)?.serverDefinitions.get();
+				if (publishedServers) {
+					this.userCache.storeServers(collectionId, { servers: publishedServers.map(McpServerDefinition.toSerialized) });
 				}
 			}
-			for (const def of nextDefinitions) {
-				nextServers.push(instantiationService.createInstance(McpServer, def.collectionDefinition, def.serverDefinition, def.collectionDefinition.scope === StorageScope.WORKSPACE ? this.workspaceCache : this.userCache));
-			}
+		}));
 
-			this._servers.set(nextServers, undefined);
-		}, 500));
+		this.userCache = this._register(_instantiationService.createInstance(McpServerMetadataCache, StorageScope.PROFILE));
+		this.workspaceCache = this._register(_instantiationService.createInstance(McpServerMetadataCache, StorageScope.WORKSPACE));
+
+		const updateThrottle = this._store.add(new RunOnceScheduler(() => this._updateCollectedServers(), 500));
 
 		// Throttle changes so that if a collection is changed, or a server is
 		// unregistered/registered, we don't stop servers unnecessarily.
 		this._register(autorun(reader => {
-			definitionsObservable.read(reader);
+			for (const collection of this._mcpRegistry.collections.read(reader)) {
+				collection.serverDefinitions.read(reader);
+			}
+			this._extensionServers.read(reader);
 			updateThrottle.schedule(500);
 		}));
 
+		const extensionCollections = this._register(new DisposableMap<string>());
 
-		const tools = this._register(new MutableDisposable());
+		_mcpExtensionPoint.setHandler((_extensions, delta) => {
```

Comment:
Added this to replace the custom extension handling. This maps contributions to collections. It also means the upsert-logic has changed, it's only insert because the collection is defined "statically"